osv2/bulk.go

```diff
@@ -0,0 +1,180 @@
+			validationResults.Add(opensearchtools.NewValidationResult("Doc ID is empty", true))
+		}
+	}
+
+	return validationResults
+}
+
+// NewBulkRequest instantiates an empty BulkRequest
+func NewBulkRequest() *BulkRequest {
+	return &BulkRequest{}
+}
+
+// Add an action to the BulkRequest.
+func (r *BulkRequest) Add(actions ...opensearchtools.BulkAction) *BulkRequest {
+	r.Actions = append(r.Actions, actions...)
+	return r
+}
+
+// WithIndex on the request
+func (r *BulkRequest) WithIndex(index string) *BulkRequest {
+	r.Index = index
+	return r
+}
+
+// ToOpenSearchJSON marshals the BulkRequest into the JSON format expected by OpenSearch.
+// Note: A BulkRequest is multi-line json with new line delimiters. It is not a singular valid json struct.
+//
+// { action1 json }
+// { action2 json }
+func (r *BulkRequest) ToOpenSearchJSON() ([]byte, error) {
+	if len(r.Actions) == 0 {
+		return nil, fmt.Errorf("bulk request requires at least one action")
+	}
+
+	bodyBuf := new(bytes.Buffer)
+	for _, op := range r.Actions {
+		jsonLines, jErr := op.MarshalJSONLines()
+		if jErr != nil {
+			return nil, jErr
+		}
+
+		for _, line := range jsonLines {
+			bodyBuf.Write(line)
+			bodyBuf.WriteRune('\n')
+		}
+	}
+
+	return bodyBuf.Bytes(), nil
+}
+
+// Do executes the BulkRequest using the provided opensearch.Client.
+// If the request is executed successfully, then a BulkResponse will be returned.
+// An error can be returned if
+//
+//   - Any Action is missing an action
+//   - The call to OpenSearch fails
+//   - The result json cannot be unmarshalled
+func (r *BulkRequest) Do(ctx context.Context, client *opensearch.Client) (*opensearchtools.OpenSearchResponse[BulkResponse], error) {
+	vrs := r.Validate()
+	if vrs.IsFatal() {
+		return nil, opensearchtools.NewValidationError(vrs)
+	}
+
+	rawBody, jErr := r.ToOpenSearchJSON()
+	if jErr != nil {
+		return nil, jErr
+	}
+
+	osResp, rErr := opensearchapi.BulkRequest{
+		Body:    bytes.NewReader(rawBody),
+		Refresh: string(r.Refresh),
+		Index:   r.Index,
+	}.Do(ctx, client)
+
+	if rErr != nil {
+		return nil, rErr
+	}
+
+	var respBuf bytes.Buffer
+	if _, err := respBuf.ReadFrom(osResp.Body); err != nil {
+		return nil, err
+	}
+
+	resp := BulkResponse{}
+
+	if err := json.Unmarshal(respBuf.Bytes(), &resp); err != nil {
+		return nil, err
+	}
+
+	return &opensearchtools.OpenSearchResponse[BulkResponse]{
+		StatusCode:        osResp.StatusCode,
+		Header:            osResp.Header,
+		Response:          resp,
+		ValidationResults: vrs,
+	}, nil
+}
+
+// BulkResponse wraps the functionality of [opensearchapi.Response] by unmarshalling the api response into
+// a slice of [bulk.ActionResponse].
+type BulkResponse struct {
+	Took   int64                            `json:"took"`
+	Errors bool                             `json:"errors"`
+	Items  []opensearchtools.ActionResponse `json:"items"`
+	Error  *Error                           `json:"error,omitempty"`
```

Comment:
I think this should just be an `Error` not a `*Error`

Reply:
The json marshaler does not omitempty structs. So in order for it to actually be omitted it an error doesn't exist, I need it to be a pointer.